File: Gate One/Lagbaja/Lagbaja.java
import java.util.Scanner;
import java.util.InputMismatchException;
import java.util.Arrays;

public class Lagbaja{

	static Scanner input = new Scanner(System.in);
	private int noOfStudents;
	private int noOfSubjects;
	private int[][] scores;
	private int[] totalPerStudent;
	private double[] averageScores;
	private int[] positions;

	public void setNoOfStudents(){
		while(true){
			try{
				System.out.print("Enter number of Students: ");
				noOfStudents = input.nextInt();
				if(noOfStudents < 0){
					System.out.println("Invalid input!!!. Enter a valid number!!!");
				} else {
					this.noOfStudents = noOfStudents;
					break;
				}
			} catch (InputMismatchException e){
				System.out.println("Invalid input. Please enter an integer.");
                        	input.next();
			}
		}
	}

	public int getNoOfStudents(){
		return noOfStudents;
	}

	public void setNoOfSubjects(){
		while(true){
			try{
				System.out.print("Enter number of Subjects: ");
				noOfSubjects = input.nextInt();

				if(noOfSubjects < 0){
				 System.out.println("Invalid input!!!. Enter a valid input.");
				} else {
					this.noOfSubjects = noOfSubjects;
					break;
				}
			} catch (InputMismatchException e){
				System.out.println("Enter a valid integer!!!");
				input.next();
			}
		}
	}

	public int getNoOfSubjects(){
		return noOfSubjects;
	}

	public int[][] collectScores() {
    		Scanner input = new Scanner(System.in);

    		scores = new int[this.noOfStudents][this.noOfSubjects];

		for (int student = 0; student < noOfStudents; student++) {
        	    System.out.println("STUDENT " + (student + 1) + ":");
        		for (int subject = 0; subject < noOfSubjects; subject++) {

				System.out.print("Subject " + (subject + 1) + ": ");
            			int score = 0;

				boolean isValid = false;

				while(!isValid){
					try{
						score = input.nextInt();
						if(score < 0 || score > 100){
							System.out.println("Invalid Score!!!. Enter a score between 0 and 100");
						} else {
							isValid = true;
						}
					} catch(InputMismatchException e){
						System.out.println("Invalid input. Please enter an integer.");
                        			input.next();
					}
				}
				scores[student][subject] = score;
			}
			System.out.println();	
            	}
		return scores;
	}

	public int[][] getScores() {
        	return scores;
    	}

	public int[] computeTotalPerStudent() {
    		totalPerStudent = new int[this.noOfStudents];

        	for (int student = 0; student < noOfStudents; student++) {
        	   int total = 0;

        		for (int subject = 0; subject < noOfSubjects; subject++) {
            			total += scores[student][subject];
        		}

        	   totalPerStudent[student] = total;
    		}
    		return totalPerStudent;
	}

	public double[] computeAveragePerStudent(){
		int[] total = computeTotalPerStudent();
		averageScores = new double[this.noOfStudents];

		double average = 0;
		for(int student = 0; student < total.length; student++){
			average = (double) total[student] / noOfSubjects;

			averageScores[student] = average;
		}
		return averageScores;
	}

	public int[] computePositions() {
    		positions = new int[this.noOfStudents];
    		Integer[] indices = new Integer[this.noOfStudents];

    		for (int index = 0; index < this.noOfStudents; index++) {
        		indices[index] = index;
    		}

    		Arrays.sort(indices, (index1, index2) -> Integer.compare(totalPerStudent[index2], totalPerStudent[index1]));

    		int currentPosition = 1;
    		int previousTotal = -1;
    		int sameRankCount = 0;

    		for (int i = 0; i < this.noOfStudents; i++) {
        		int studentIndex = indices[i];
        		int currentTotal = totalPerStudent[studentIndex];

        		if (currentTotal == previousTotal) {
            			positions[studentIndex] = currentPosition;
            			sameRankCount++;
        		} else {
            			currentPosition = i + 1;
            			positions[studentIndex] = currentPosition;
            			previousTotal = currentTotal;
            			sameRankCount = 1;
        		}
    		}
    			return positions;
	}

    public void displayResults() {
        System.out.println("\nResults:");

        System.out.printf("%-10s", "Student");
        for (int subject = 0; subject < noOfSubjects; subject++) {
            System.out.printf("%-10s", "Subject" + (subject + 1));
        }
        System.out.printf("%-10s%-10s%-10s\n", "Total", "Average", "Position");

        for (int student = 0; student < noOfStudents; student++) {
            System.out.printf("%-10s", "Student" + (student + 1));
            for (int subject = 0; subject < noOfSubjects; subject++) {
                System.out.printf("%-10d", scores[student][subject]);
            }
            System.out.printf("%-10d%-10.2f%-10d\n", totalPerStudent[student], averageScores[student], positions[student]);
        }
    }
}

File: Gate One/Lagbaja/lagbaja.py
def set_no_of_students():
    while True:
        try:
            no_of_students = int(input("Enter number of Students: "))
            if no_of_students <= 0:
                print("Invalid input! Enter a positive number.")
            else:
                return no_of_students
        except ValueError:
            print("Invalid input. Please enter an integer.")

def set_no_of_subjects():
    while True:
        try:
            no_of_subjects = int(input("Enter number of Subjects: "))
            if no_of_subjects <= 0:
                print("Invalid input! Enter a positive number.")
            else:
                return no_of_subjects
        except ValueError:
            print("Invalid input. Please enter an integer.")

def collect_scores(no_of_students, no_of_subjects):
    scores = []
    for student in range(no_of_students):
        student_scores = []
        print(f"STUDENT {student + 1}:")
        for subject in range(no_of_subjects):
            while True:
                try:
                    score = int(input(f"Subject {subject + 1}: "))
                    if 0 <= score <= 100:
                        student_scores.append(score)
                        break
                    else:
                        print("Invalid Score! Enter a score between 0 and 100.")
                except ValueError:
                    print("Invalid input. Please enter an integer.")
        scores.append(student_scores)
    return scores

def compute_total_per_student(scores):
    return [sum(student_scores) for student_scores in scores]

def compute_average_per_student(total_per_student, no_of_subjects):
    return [total / no_of_subjects for total in total_per_student]

def compute_positions(total_per_student):
    indices = list(range(len(total_per_student)))
    indices.sort(key=lambda x: total_per_student[x], reverse=True)

    positions = [0] * len(total_per_student)
    current_position = 1
    previous_total = -1
    for index in indices:
        if total_per_student[index] == previous_total:
            positions[index] = current_position
        else:
            current_position = index + 1
            positions[index] = current_position
            previous_total = total_per_student[index]
    return positions

def display_results(scores, total_per_student, average_scores, positions):
    print("\nResults:")
    
    header = ["Student"] + [f"Subject{subject + 1}" for subject in range(len(scores[0]))] + ["Total", "Average", "Position"]
    print("{:<10}".format(header[0]), end='')
    for item in header[1:]:
        print("{:<10}".format(item), end='')
    print()
    
    for student, student_scores in enumerate(scores):
        print(f"Student{student + 1:<10}", end='')
        for score in student_scores:
            print("{:<10}".format(score), end='')
        print("{:<10}".format(total_per_student[student]), end='')
        print("{:<10.2f}".format(average_scores[student]), end='')
        print("{:<10}".format(positions[student]))


no_of_students = set_no_of_students()
no_of_subjects = set_no_of_subjects()
print(f"No of Students: {no_of_students}, No of Subjects: {no_of_subjects}")
scores = collect_scores(no_of_students, no_of_subjects)

total_per_student = compute_total_per_student(scores)
average_scores = compute_average_per_student(total_per_student, no_of_subjects)
positions = compute_positions(total_per_student)
print()
display_results(scores, total_per_student, average_scores, positions)